man/listall.man
.\" 3.2
.\"  /*******************************************************************\
.\"   * Copyright (C) 1995 Lars J Aas <larsa@colargol.stud.idb.hist.no> *
.\"   * based on documentation by C J Needham <col@imdb.com> 1995,      *
.\"   * permission is granted by the authors to freely distribute       *
.\"   *                        providing no fee of any kind is charged. *
.\"  \*******************************************************************/
.\"
.TH LISTALL 1 "RELEASEDATE" " " "The Internet Movie Database vVERSIONNUMBER"
.SH NAME
listall \- for all persons involved in a movie, list all the movies they have been involved with.
.SH SYNOPSYS
.B
listall -t "title" [<options>]
.SH DESCRIPTION
.B listall
takes a movie title as an argument and for each person involved in 
that movie, lists all the other movies they have worked in.
.SH OPTIONS
.TP
.B \-yr
Add year of release to titles and sort chronologically.
.TP
.B \-yru
Add year of release, but leave in alphabetical order.
.TP
.B \-mrr
Add movie ratings report information.
.TP
.B \-smrr
Add movie ratings report and sort by average ratings.
.TP
.B \-vmrr
Add movie ratings report and sort by votes.
.TP
.B \-chr
Add character names where available.
.TP
.B \-bio
Add any biography data.
.TP
.B \-aka
Add any alternative titles (and check if title is an aka title).
.TP
.B \-full
Combine
.BR \-aka ,
.BR \-bio ,
.B \-chr
and
.BR \-yr .
.TP
.B \-i
Case sensitive search.
.TP
.B \-m
Restrict search to movies only.
.SH EXAMPLE
listall -t "Abyss, The" -full


docs/ChangeLog
-------Version 3.4 Released---------------------

Sun Mar 29 22:40:37 BST 1998

 * biographies list offset fields increased to 4 bytes following overflow
 * various constants updated

-------Version 3.3c Released---------------------

Sat Apr 26 06:51:19 BST 1997

 * added support for crew completion list
 * added support for distributors list
 * adapted mkdbs for more robust processing of filmography lists
 * added support for new business list tag

-------Version 3.3b Released---------------------

Sat Mar 15 08:09:50 GMT 1997

 * added support for ISO-Latin-1

-------Version 3.3 Released---------------------

Wed Oct 16 22:28:54 BST 1996

 * added support for AKA attributes
 * added support for writers ordering
 * moved misc section to after case in title output
 * added support for business list
 * added support for laserdisc list
 * fixed tagline bug
 * diplays URL of web page for name/title

-------Version 3.2h Released---------------------

Tue Oct  8 14:24:21 BST 1996

 * added support for special effects companies list
 * updated various system constants
 * added new movie link types 'edited into' & 'edited from'
 * added extra checks to avoid title/attribute key file corruption in mkdbs.c

-------Version 3.2f Released---------------------

Mon Jun 24 22:04:46 GMT 1996

 * added support for language list

-------Version 3.2e Released---------------------

Sat Mar 16 08:09:39 GMT 1996

 * added support for alternate versions list
 * added bio keys IT & AT

-------Version 3.2d Released---------------------

Tue Aug 15 23:23:47 GMT 1995

 * switched TitleID to unsigned long to handle >65K titles
 * fixed bug in display of literature and genre data
 * added bio key WN

-------Version 3.2c Released---------------------

Sun May 14 23:00:14 GMT 1995

 * raised MAXFILMOGRAPHIES constant to handle larger no. of entries

-------Version 3.2 Released---------------------

Sun Jan 29 22:40:08 GMT 1995

 * added full support for title searching without year/TV/V info

Tue Jan  3 23:00:45 GMT 1995

 * replaced 'biodefs' file
 * fixed bug when looking for common titles with people using -name

Mon Jan  2 21:52:08 GMT 1995

 * log file support
 * better separation of search/display code

Wed Dec 14 21:43:57 GMT 1994

 * fixed bug in 'title' for miscellaneous section
 * added support for the movie links list

Tue Dec 13 22:06:44 GMT 1994
 
 * noted when cast in credits output
 * fixed name ordering problem for names with (I)/(II) marker
 * display Actor/Actress in list -name where appropriate

Wed Dec  7 21:19:19 GMT 1994

 * added support for tag lines & complete cast list
 * put spaces around "." in cast results
 * standardized number of dashes printed out in 'title' / 'list'

Mon Dec  5 21:19:42 GMT 1994

 * multi-country productions displayed on same line

Wed Nov 30 18:14:16 GMT 1994

 * switched to new IMDB license
 * updated 'lfetch' to get new lists
 * fixed "No plot summaries message"
 * simplified MRR processing in mkdb
 * added support for locations & release date lists
 * switched 'Reviews' marker for clarity
 * added support for technical & literature lists

Mon Oct 31 21:42:03 GMT 1994
 
-------Version 3.1 Released---------------------

Mon Oct 17 19:04:26 GMT 1994
 
 * 3.1 code frozen
 * minor bug fixes with lindex

Sun Oct  9 18:37:14 GMT 1994

 * added support for countries list

Tue Oct  4 21:27:53 GMT 1994

 * added support for cast ordering info

Sun Oct  2 23:02:59 GMT 1994

 * added support for color info and sound mix lists
 * added constraint options to 'list'
 * standardized list options

Wed Sep 28 22:22:56 GMT 1994

 * re-ordered title info printing plus separated genres list
 * print genres on one line

Mon Sep 26 21:53:47 GMT 1994

 * added support for production companies list

Sun Sep 25 22:02:07 GMT 1994

 * added constraint options to 'lindex'

Tue Sep  6 20:09:52 GMT 1994

 * added support for certificates & genres list
 * fixed mkballot bug with -ar option

Tue Aug 23 20:08:04 GMT 1994

 * added support for miscellaneous filmographies list

Tue May 24 19:13:47 GMT 1994

 * moviedb-3.0 completed

Thu May 19 22:35:12 GMT 1994

 * switched to merged character names scheme

Tue May 17 21:09:45 GMT 1994

 * new 'mkdb' eliminating need for large array to hold name keys
 * removed duplicate aka titles in 'title -s -aka' queries

Wed May 11 22:11:59 GMT 1994

 * made upperCase a macro for increased speed
 * switched to Phil's new Makefiles

Wed May  4 17:46:08 GMT 1994

 * mapped case of titles/names to match key files
 * added support for running times list
 * lguide error messages

Tue May  3 21:38:30 GMT 1994
 
 * added support for Phil's imoviedb package
 * expanded compression support

Thu Apr 28 19:51:52 GMT 1994

 * finished rewrite to use new database representation

-------Version 2.9 Released---------------------

Wed Mar 16 21:39:40 GMT 1994

 * added support for producers list
 * added support for soundtracks list
 * added support for generic aka lists

Tue Mar 15 23:47:46 GMT 1994

 * replaced -crazy and -triv options with -trivia (also includes goofs & quotes)
 * added support for quotes list
 * added support for goofs list
 * handled -full error messages correctly
 * correct conditions for printing of missing plot summaries message
 * fixed lindex to search all lists when no list options given
 * -mrr no longer implies -m

Mon Mar 14 21:48:21 GMT 1994

 * new version of check-for-program
 * passed $MAKE variable to make subprocesses
 * lfetch modified to match full filenames
 * added 'keepchars' option for CLEANLISTS

-------Version 2.8 Released---------------------

Wed Dec  1 21:58:21 GMT 1993

 * man pages added to package
 * LaTeX documentation added to package

Sat Nov 27 20:53:40 GMT 1993

 * 2.8 code finished!

Thu Nov 25 18:17:00 GMT 1993

 * ported lseen to C

Tue Nov 23 19:16:48 GMT 1993

 * added gendb -votes option

Mon Nov 22 23:54:58 GMT 1993

 * ported lformat to C
 * removed yrf.c file

Wed Nov 17 22:36:44 GMT 1993


src/listall.c
/*============================================================================ 
 * 
 *  Program: listall.c  
 * 
 *  Version: 3.7
 * 
 *  Purpose: search list databases 
 * 
 *  Author:  C J Needham <col@imdb.com>     
 * 
 *  Copyright (c) 1996-1999 The Internet Movie Database Inc.
 * 
 *  Permission is granted by the copyright holder to distribute this program
 *  is source form only, providing this notice remains intact, and no fee
 *  of any kind is charged. This software is provided as is and there are no 
 *  warranties, expressed or implied.
 *  
 *  Options: 
 *    -t <title>  search for details on title 
 *    -aka        check for alternative title 
 *    
 *    -yr         display year of release 
 *    -yru        display year of release, unsorted 
 * 
 *    -mrr        display movie ratings report info 
 *    -smrr       display movie ratings report info, sorted by ratings 
 *    -vmrr       display movie ratings report info, sorted by votes 
 * 
 *    -chr        add character names 
 *    -bio        add biography data
 *    -full       combination of -yr -aka -chr -bio
 *
 *============================================================================ 
 */ 
 
#include <stdio.h> 
#include <string.h> 
#include <stdlib.h> 
#include "log.h" 
#include "moviedb.h" 
#include "dbutils.h" 
#include "titlesearch.h" 
#include "filmography.h" 
#include "display.h" 
 
#define LISTALL_USAGE1 "usage: listall -t <title> [-m] [-mrr|-smrr|-vmrr] [-yr|-yru]" 
#define LISTALL_USAGE2 "               [-aka] [-bio] [-chr] [-full]" 
 
int main ( int argc, char **argv ) 
{ 
  struct titleSearchRec *trec = NULL ; 
  struct titleSearchOptRec toptions = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 } ;
  struct nameSearchRec  *nchain = NULL ; 
  struct nameSearchOptRec noptions = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 } ;
  int   err = FALSE ;
  int   i, j ;
  int   trivopts [ NO_OF_TRIV_LISTS ] ;
 
  logProgram ( argv [ 0 ] ) ; 
  toptions . yearMatch = TRUE ;

  if ( argc == 1 ) 
    err = TRUE ; 

  for ( i = 0 ; i < NO_OF_TRIV_LISTS ; i++ )
    trivopts [ i ] = FALSE ;
 
  for ( i = 1 ; i < argc; i++ ) 
    if ( strcmp ( "-t", argv[i] ) == 0 ) 
      if ( ++i < argc ) 
        trec = addTitleSearchRec ( trec, argv [ i ] ) ; 
      else 
        err = TRUE ; 
    else if ( strcmp ( "-m", argv [ i ] ) == 0 ) 
      noptions . mvsonly = TRUE ; 
    else if ( strcmp ( "-aka", argv [ i ] ) == 0 ) 
    {
      noptions . akaopt = TRUE ; 
      toptions . akaopt = TRUE ; 
    }
    else if ( strcmp ( "-exact", argv [ i ] ) == 0 )
      toptions . yearMatch = FALSE ;
    else if ( strcmp ( "-mrr", argv [ i ] ) == 0 ) 
      noptions . mrropt = MRR ; 
    else if ( strcmp ( "-smrr", argv [ i ] ) == 0 ) 
      noptions . mrropt = SMRR ; 
    else if ( strcmp ( "-vmrr", argv [ i ] ) == 0 ) 
      noptions . mrropt = VMRR ; 
    else if ( strcmp ( "-yr", argv [ i ] ) == 0 ) 
      noptions . yropt = YR ; 
    else if ( strcmp ( "-yru", argv [ i ] ) == 0 ) 
      noptions . yropt = YRU ; 
    else if ( strcmp ( "-chr", argv [ i ] ) == 0 ) 
      noptions . chopt = TRUE ; 
    else if ( strcmp ( "-full", argv [ i ] ) == 0 ) 
    {
      noptions . biopt = TRUE ;
      noptions . chopt = TRUE ;
      noptions . akaopt = TRUE ;
      noptions . yropt = YR ; 
      toptions . akaopt = TRUE ;
      toptions . yropt = TRUE ;
      toptions . mrropt = TRUE ;
      toptions . plotopt = TRUE ;
      toptions . businessopt = TRUE ;
      toptions . ldopt = TRUE ;
      toptions . chopt = TRUE ;
      toptions . litopt = TRUE ;
      toptions . linkopt = TRUE ;
      for ( j = 0 ; j < NO_OF_TRIV_LISTS ; j++ )
        trivopts [ j ] = TRUE ;
    }
    else
    { 
       (void) fprintf ( stderr, "listall: unrecognised option %s\n", argv[i] ) ; 
       err = TRUE ; 
    } 
 
  if ( err || trec == NULL ) 
    moviedbUsage ( LISTALL_USAGE1, LISTALL_USAGE2, NULL, NULL, NULL, NULL ) ;
 
  if ( trec -> next != NULL ) 
    moviedbError ( "listall: only one title allowed" ) ; 
 
  addTitleChainOpts ( trec, toptions ) ;
  addTitleChainTrivOpts ( trec, trivopts ) ;
  processTitleSearch ( trec ) ;
  if ( trec -> next != NULL )
    displayListallTitleMatches ( trec ) ;
  else
  {
    nchain = makeNameChain ( trec ) ; 
    if ( nchain != NULL )
    {
      addNameChainOpts ( nchain, noptions ) ;
      processFilmographySearch ( nchain ) ;
      sortListResults ( nchain ) ;
      displayTitleSearchResults ( trec, TRUE ) ;
      displayNameSearchResults ( nchain, TRUE ) ;
      freeNameSearchChain ( nchain ) ; 
    }
  }
  freeTitleChain ( trec ) ; 
  return ( 0 ) ; 
} 
